plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+  }
+  if (os_resp_reader) {
+    TSIOBufferReaderFree(os_resp_reader);
+  }
+  if (os_resp_buf) {
+    TSIOBufferDestroy(os_resp_buf);
+  }
+  if (icap_resp_reader) {
+    TSIOBufferReaderFree(icap_resp_reader);
+  }
+  if (icap_resp_buf) {
+    TSIOBufferDestroy(icap_resp_buf);
+  }
+  if (output_reader) {
+    TSIOBufferReaderFree(output_reader);
+  }
+  if (output_buf) {
+    TSIOBufferDestroy(output_buf);
+  }
+  if (pending_action) {
+    TSActionCancel(pending_action);
+  }
+  if (icap_vc) {
+    TSVConnAbort(icap_vc, 1);
+  }
+}
+
+/*
+ * get_port
+ * Description: Return the port of a sockaddr
+ */
+uint16_t
+get_port(sockaddr const *s_sockaddr)
+{
+  switch (s_sockaddr->sa_family) {
+  case AF_INET: {
+    const struct sockaddr_in *s_sockaddr_in = reinterpret_cast<const struct sockaddr_in *>(s_sockaddr);
+    return ntohs(s_sockaddr_in->sin_port);
+  } break;
+  case AF_INET6: {
+    const struct sockaddr_in6 *s_sockaddr_in6 = reinterpret_cast<const struct sockaddr_in6 *>(s_sockaddr);
+    return ntohs(s_sockaddr_in6->sin6_port);
+  } break;
+  default:
+    return 0;
+    break;
+  }
+}
+
+/*
+ * setup_icap_status_header (Used only in debug-mode)
+ * Description: This function is called to add a customized header
+ *              indicating ICAP server status for logging.
+ */
+static void
+setup_icap_status_header(TransformData *data, const char *header, const char *value)
+{
+  TSMBuffer bufp;
+  TSMLoc resp_loc, field_loc;
+
+  if (TSHttpTxnTransformRespGet(data->txn, &bufp, &resp_loc) != TS_SUCCESS) {
+    TSError("[%s] Couldn't retrieve transform response header", PLUGIN_NAME);
+    return;
+  }
+
+  if (TSMimeHdrFieldCreate(bufp, resp_loc, &field_loc) != TS_SUCCESS) {
+    TSError("[%s] Unable to create field", PLUGIN_NAME);
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, resp_loc);
+    return;
+  }
+
+  TSMimeHdrFieldNameSet(bufp, resp_loc, field_loc, header, strlen(header));
+  TSMimeHdrFieldValueStringInsert(bufp, resp_loc, field_loc, 0, value, strlen(value));
+  TSMimeHdrFieldAppend(bufp, resp_loc, field_loc);
+
+  TSHandleMLocRelease(bufp, resp_loc, field_loc);
+  TSHandleMLocRelease(bufp, TS_NULL_MLOC, resp_loc);
+}
+
+/*
+ * handle_invalid_icap_behavior
+ * Description: This function is called when abnormal behavior of ICAP server,
+ *              for instance, unsuccessful connection, is detected.
+ */
+static int
+handle_invalid_icap_behavior(TSCont contp, TransformData *data, const char *msg)
+{
+  if (data->icap_vc) {
+    TSVConnAbort(data->icap_vc, 1);
+    data->icap_vc  = NULL;
+    data->icap_vio = NULL;
+  }
+  TSStatIntIncrement(total_icap_invalid, 1);
+  TSDebug(PLUGIN_NAME, "\n%s\n", data->icap_header.c_str());
+  data->err_msg = std::string(msg);
+  /* Signal the upstream vconn if still exists to stop sending data */
+  TSVIO write_vio = TSVConnWriteVIOGet(contp);
+  if (TSVIOBufferGet(write_vio)) {
+    TSContCall(TSVIOContGet(write_vio), TS_EVENT_VCONN_WRITE_COMPLETE, write_vio);
```

Comment:
Hmm TSContCall() does a release assert if the continuation has a mutex and try lock on the mutex fails.